presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3FileSystem.java

```diff
@@ -298,14 +301,95 @@ public FSDataOutputStream append(Path f, int bufferSize, Progressable progress)
 
     @Override
     public boolean rename(Path src, Path dst)
+        throws IOException
     {
-        throw new UnsupportedOperationException("rename");
+        FileStatus srcStatus;
+        try {
+            srcStatus = getFileStatus(src);
+        }
+        catch (FileNotFoundException e) {
+            log.debug("Try to rename file: %s, But file does not exist", src.toString());
+            return false;
+        }
+
+        String srcKey = keyFromPath(src);
+        String dstKey = keyFromPath(dst);
+
+        FileStatus dstStatus;
+        try {
+            dstStatus = getFileStatus(dst);
+            if (dstStatus.isDirectory()) {
+                // move source under destination directory
+                dstKey = keyFromPath(new Path(dst, src.getName()));
+            }
+            else if (dstStatus.isFile()) {
+                // not copy a file to an existing file
+                return srcKey.equals(dstKey);
+            }
+        }
+        catch (FileNotFoundException e) {
+            // destination does not exist
+        }
+
+        if (srcKey.equals(dstKey)) {
+            return true;
+        }
+
+        if (srcStatus.isFile()) {
+            s3.copyObject(uri.getHost(), srcKey, uri.getHost(), dstKey);
+            delete(src, true);
+        }
+
+        if (srcStatus.isDirectory()) {
+            FileStatus[] files = listStatus(src);
+            for (FileStatus file : files) {
+                rename(file.getPath(), new Path(dst, file.getPath().getName()));
```

Comment:
This broken, because it uses `dst` not `dstKey`, so it misses the adjustment from above. I found this when I added some tests :)